packages/langium/src/grammar/references/langium-grammar-references.ts

```diff
@@ -0,0 +1,248 @@
+/******************************************************************************
+* Copyright 2021 TypeFox GmbH
+* This program and the accompanying materials are made available under the
+* terms of the MIT License, which is available in the project root.
+******************************************************************************/
+
+import { DefaultReferences, FindReferencesOptions } from '../../references/references';
+import { LangiumServices } from '../../services';
+import { AstNode, CstNode, Reference } from '../../syntax-tree';
+import { extractAssignments, findNameNode, getContainerOfType, getDocument, streamAst, streamReferences } from '../../utils/ast-util';
+import { findLeafNodeAtOffset, findRelevantNode, toDocumentSegment } from '../../utils/cst-util';
+import { stream, Stream } from '../../utils/stream';
+import { equalURI } from '../../utils/uri-utils';
+import { ReferenceDescription } from '../../workspace/ast-descriptions';
+import { LangiumDocuments } from '../../workspace/documents';
+import { Action, Assignment, Interface, isAction, isAssignment, isGroup, isInterface, isParserRule, isType, isTypeAttribute, ParserRule, Type, TypeAttribute } from '../generated/ast';
+import { findNodeForFeature } from '../grammar-util';
+import { collectChildrenTypes, collectSuperTypes } from '../type-system/types-util';
+
+export class LangiumGrammarReferences extends DefaultReferences {
+    protected readonly documents: LangiumDocuments;
+
+    constructor(services: LangiumServices) {
+        super(services);
+        this.documents = services.shared.workspace.LangiumDocuments;
+    }
+
+    findDeclaration(sourceCstNode: CstNode): CstNode | undefined {
+        const nodeElem = findRelevantNode(sourceCstNode);
+        if (isAssignment(nodeElem)) {
+            return this.findAssignmentDeclaration(nodeElem);
+        }
+        return super.findDeclaration(sourceCstNode);
+    }
+
+    findReferences(target: AstNode, options: FindReferencesOptions): Stream<ReferenceDescription> {
+        if (options.onlyLocal) {
+            return this.findLocalReferences(target, options.includeDeclaration);
+        } else {
+            return this.findGlobalReferences(target, options.includeDeclaration);
+        }
```

Comment:
Suggestion: As mentioned above, the default implementation should respect the `onlyLocal` property already and ideally provide a template method to modify the behavior of both execution flows.